pkg/provision/engine.go

```diff
@@ -71,16 +77,25 @@ func New(opts EngineOps) *Engine {
 		decomissioners: opts.Decomissioners,
 		signer:         opts.Signer,
 		statser:        opts.Statser,
+		zbusCl:         opts.ZbusCl,
 	}
 }
 
 // Run starts reader reservation from the Source and handle them
 func (e *Engine) Run(ctx context.Context) error {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
 	cReservation := e.source.Reservations(ctx)
 
+	// run a cron task that will fire the cleanup at midnight
+	cleanUp := make(chan struct{}, 2)
+	c := cron.New()
+	_, err := c.AddFunc("@midnight", func() {
```

Comment:
It feels like there is a race condition here, no?

Reply:
race condition on what ?